### PR TITLE
Permit attributes on 'if' expressions

### LINKED ISSUE #68618: Allow `#[attr] if` to be passed to proc macros
The following code:

```rust
fn main() {
    #[attr] if let Some(_) = Ok(true) {}
}
```

produces this error:

```
error: attributes are not yet allowed on `if` expressions
 --> src/main.rs:2:5
  |
2 |     #[attr] if let Some(_) = Ok(true) {}
  |     ^^^^^^^

error: aborting due to previous error
```

Proc macros like [`pin-project`](https://github.com/taiki-e/pin-project) currently use a trick to work around attributes on expressions being unstable: The entire function is annotated with a 'wrapper' attribute, which goes through and manually parses expression attributes. For example:

```rust
#[my_attr]
fn foo() {
    #[my_attr] let a = 25;
}
```

In the example above, the `#[my_attr] let a = 25;` will be processed by the outer `#[my_attr] fn foo()` implementation, which will strip away the expression-based `#[my_attr]` before emitting the final `TokenStream`.

Unfortunately, the current 'attributes are not yet allowed on `if` expressions' check is done during parsing:

https://github.com/rust-lang/rust/blob/ac2f3fa41ac5ae8425b959f955bb7433b7c57aea/src/librustc_parse/parser/expr.rs#L679

While the proc-macro itself can parse the attribute on the `if let` without any errors, compilation will still fail due to the parsing error that was emitted *before* any proc-macros were run.

It would be extremely useful if this check were to be moved until **after** parsing, so that proc macros had a chance to remove attributes from `if` expressions.

### PR DESCRIPTION
**Pull request by Aaron1011**
Previously, attributes on 'if' expressions (e.g. `#[attr] if true {}`)
were disallowed during parsing. This made it impossible for macros to
perform any custom handling of such attributes (e.g. stripping them
away), since a compilation error would be emitted before they ever had a
chance to run.

This PR permits attributes on 'if' expressions ('if-attrs' from here on).
Both built-in attributes (e.g. `#[allow]`, `#[cfg]`) and proc-macro attributes are supported.

We still do *not* accept attributes on 'other parts' of an if-else
chain. That is, the following code snippet still fails to parse:

```rust
if true {} #[attr] else if false {} else #[attr] if false {} #[attr]
else {}
```

Closes https://github.com/rust-lang/rust/issues/68618

### CODE REVIEW

File: src/test/ui/parser/attr-stmt-expr-attr-bad.rs [hunk 1]
```diff
@@ -51,14 +49,11 @@ fn main() {}
...
 //~^ ERROR an inner attribute is not permitted in this context
 #[cfg(FALSE)] fn e() { let _ = if 0 {} else #[attr] if 0 {}; }
-//~^ ERROR attributes are not yet allowed on `if` expressions
-//~| ERROR expected `{`, found `#`
+//~^ ERROR expected `{`, found `#`
 #[cfg(FALSE)] fn e() { let _ = if 0 {} else if 0 #[attr] {}; }
 //~^ ERROR expected `{`, found `#`
 #[cfg(FALSE)] fn e() { let _ = if 0 {} else if 0 {#![attr]}; }
 //~^ ERROR an inner attribute is not permitted in this context
-#[cfg(FALSE)] fn e() { let _ = #[attr] if let _ = 0 {}; }
```
Comment (Centril) [inline src/test/ui/parser/attr-stmt-expr-attr-bad.rs]: Let's extend the pretty test with this case (with & without the inner `let`) extended with an `else` branch.

File: src/test/ui/parser/recovery-attr-on-if.rs [hunk 2]
```diff
@@ -1,9 +1,7 @@
 fn main() {
     #[attr] if true {};
```
Comment (Centril) [inline src/test/ui/parser/recovery-attr-on-if.rs]: This test isn't testing recovery anymore and should be removed.

**Review by Centril — COMMENTED** [review]
Looks good; I'd like to see some more tests though:
- Interactions with `stmt_expr_attributes`, specifically demonstrating that `let _ = #[allow(...)] if ...;` is feature gated.
- Testing that `let _ = #[cfg(FALSE)] if true {};` results in "error: removing an expression is not supported in this position".
- Testing that `#[cfg(FALSE)]` and `cfg_attr` perform normally when attached onto `if ...` as a statement. (e.g. provoke a type error inside the block in the former case and a lint set to `deny` in the latter case, and ensure that the test is check-pass).

These mostly test aspects of statements and expressions themselves, but I'd like to be sure that there's nothing out of the ordinary here.

**Comment by Centril** [comment]
From https://github.com/rust-lang/rust/pull/68658#issuecomment-585977456:

> We discussed this in today's language team meeting (2020-02-13).
>
> Those present, in particular myself and @joshtriplett, felt that of the options available (1. keep the status quo, 2. accept only syntactically, 3. accept semantically also), that the 3rd option would be the simplest, and least complex approach.
>
> Specifically, we would like to allow `#[attr]` on `if` both syntactically (in the parser) and semantically (expansion / conditional compilation, subject to existing feature gates for `stmt_expr_attributes`, e.g. `#[allow()] match 0 { _ => {} };` (ungated) vs. `let _ = #[allow()] match 0 { _ => {} };` (gated)).
>
> In reaching this conclusion, we did consult [RFC 16](https://github.com/rust-lang/rfcs/blob/master/text/0016-more-attributes.md#if) regarding attributes on `if` expressions. We noted in particular that neither of `else #[cfg(not(bar))] if cond2` or `else #[cfg(not(baz))] {`. Moreover, as @petrochenkov notes in https://github.com/rust-lang/rust/pull/68658#issuecomment-581074552:
>
>> It's pretty established now that macros and attributes work with expressions/statements and not arbitrary syntactic fragments, so the choice in favor of the alternative 2 from https://github.com/rust-lang/rfcs/blob/master/text/0016-more-attributes.md#if is unlikely.
> 
> We also noted that while many users interpret code in a linear / textual fashion as opposed to as a tree-like structure, we still believe that the semantics of e.g. `#[allow(...)] if ...` is something that can be learned as necessary.

Based on this, let's:

@rfcbot merge

**Comment by joshtriplett** [comment]
I'd like to request one more test, please:

```rust
let x = 1;

#[cfg(FALSE)]
if false {
    x = 2;
} else if true {
    x = 3;
} else {
    x = 4;
}
assert_eq!(x, 1);
```

This makes sure that the attribute gets applied to the entire chained if/else-if/else construct, and not just part of it.

**Comment by Aaron1011** [comment]
@Centril: I've added the additional tests you requested, and removed the parser recovery test.

@joshtriplett: I've added the test you suggested.

**Review by Centril — APPROVED** [review]
r=me when FCP is done

**Comment by scottmcm** [comment]
@rfcbot reviewed

I don't have a complaint about this situation in particular, but do wonder about philosophy around attributes here.  @Centril mentioned in the meeting that it's expressions in certain context that can allow attributes; I'd like a page in the reference or something about _why_ particular things can or cannot have attributes.  For example, the following currently doesn't work:
```rust
let x = #[cfg(true)] 4 #[cfg(not(true))] 5;
```
But in some sense there isn't a reason that it _couldn't be made to work_, since either way is a valid expression for the RHS of the binding.

So I'd just like a definition somewhere about why it's the right thing to do on an `if` expression, but not on a literal expression.  (Sorry if it already exists somewhere, but I didn't see an answer in the OP, in the RFC proposal comment, or in skimming the thread.)

**Comment by eddyb** [comment]
@scottmcm There is one interesting distinction: expressions `if` are allowed to be expression statements without a `;`.
AFAIK, the shortest version of your example that does work is:
```rust
let x = {
    #[cfg(foo)] { 4 }
    #[cfg(not(foo))] { 5 }
};
```

This is because this parses even if you remove the `#[cfg]`s (and results in a `mismatched types` error, as it expects `{ 4 }` to have the type `()`).

**Comment by Centril** [comment]
The reason why e.g. `#[cfg(stuff))] let x = 0;` works but e.g. `let x = #[cfg(true)] 4 #[cfg(not(true))] 5;` doesn't can I think be summarized with "because monoids". In the former context, we have a list of statements, and lists have as their identity element the empty list. In the latter context, we have an expression to the right of `=`. Expressions do not have an identity element, so what do we do if both `4` and `5` are removed? Similarly, what do we do if both are kept? Remember that the conditional compilation engine does not perform any form of "there will only be one remaining" analysis after stripping is done. (This is ignoring the fact that the parser would not allow `Attrs* 4 Attrs* 5`, so we never reach conditional compilation to begin with.)

**Pull request merged**